server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -480,87 +483,125 @@ public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request, Action
 
         clusterService.submitStateUpdateTask(
             "add-index-block-[" + request.getBlock().name + "]-" + Arrays.toString(concreteIndices),
-            new ClusterStateUpdateTask(Priority.URGENT, request.masterNodeTimeout()) {
+            new AddBlocksTask(request, listener),
+            ClusterStateTaskConfig.build(Priority.URGENT, request.masterNodeTimeout()),
+            addBlocksExecutor
+        );
+    }
 
-                private Map<Index, ClusterBlock> blockedIndices;
+    private class AddBlocksExecutor implements ClusterStateTaskExecutor<AddBlocksTask> {
+
+        @Override
+        public ClusterState execute(ClusterState currentState, List<TaskContext<AddBlocksTask>> taskContexts) throws Exception {
+            ClusterState state = currentState;
 
-                @Override
-                public ClusterState execute(final ClusterState currentState) {
-                    final Tuple<ClusterState, Map<Index, ClusterBlock>> tup = addIndexBlock(
-                        concreteIndices,
+            for (final var taskContext : taskContexts) {
+                try {
+                    final var task = taskContext.getTask();
+                    final Tuple<ClusterState, Map<Index, ClusterBlock>> blockResult = addIndexBlock(
+                        task.request.indices(),
                         currentState,
```

Comment:
Oops

```suggestion
                        state,
```

Otherwise we're not accumulating updates. Could we have a test to catch this?

Reply:
OOF.

Great catch, how embarrassing. I'll get a test in place and fix it. Thanks for catching this!

Reply:
How does f90215b79778fdf63c365eb38beecfb5f7a657a2 look to you, @DaveCTurner?